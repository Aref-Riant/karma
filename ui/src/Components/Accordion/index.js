import React from "react";

import Collapsible from "react-collapsible";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";

const Trigger = ({ text, isOpen }) => (
  <div className="d-flex flex-row justify-content-between">
    <div>{text}</div>
    <div>
      <FontAwesomeIcon
        icon={isOpen ? faChevronDown : faChevronUp}
        className="text-muted"
      />
    </div>
  </div>
);

const Accordion = ({ text, content, extraProps }) => (
  <Collapsible
    triggerTagName="div"
    transitionTime={50}
    overflowWhenOpen="visible"
    className="card"
    openedClassName="card"
    triggerClassName="card-header cursor-pointer border-bottom-0"
    triggerOpenedClassName="card-header cursor-pointer"
    contentOuterClassName="collapse show"
    contentInnerClassName="card-body my-2"
    {...extraProps}
    trigger={<Trigger text={text} isOpen={false} />}
    triggerWhenOpen={<Trigger text={text} isOpen={true} />}
  >
    {content}
  </Collapsible>
);

export { Accordion };
