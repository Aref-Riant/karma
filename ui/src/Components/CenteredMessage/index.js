import React from "react";

import { MountFade } from "Components/Animations/MountFade";

const CenteredMessage = ({ children, className }) => (
  <h1
    className={`${
      className ? className : "display-1 text-placeholder"
    } screen-center`}
  >
    <MountFade in={true}>{children}</MountFade>
  </h1>
);

export { CenteredMessage };
