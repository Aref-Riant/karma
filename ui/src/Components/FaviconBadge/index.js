import React, { Component } from "react";
import PropTypes from "prop-types";

import { observer } from "mobx-react";

import Favico from "favico.js";

import { AlertStore } from "Stores/AlertStore";

const FaviconBadge = observer(
  class FaviconBadge extends Component {
    static propTypes = {
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
    };

    constructor(props) {
      super(props);

      this.favicon = new Favico({
        animation: "none",
        position: "down",
        bgColor: "#e74c3c",
        textColor: "#fff",
        fontStyle: "lighter",
      });
    }

    updateBadge = () => {
      const { alertStore } = this.props;

      if (alertStore.status.error !== null) {
        this.favicon.badge("?");
      } else {
        this.favicon.badge(alertStore.info.totalAlerts);
      }
    };

    componentDidMount() {
      this.updateBadge();
    }

    componentDidUpdate() {
      this.updateBadge();
    }

    render() {
      const { alertStore } = this.props;

      return (
        <span
          data-total-alerts={alertStore.info.totalAlerts}
          data-status-error={alertStore.status.error}
        />
      );
    }
  }
);

export { FaviconBadge };
