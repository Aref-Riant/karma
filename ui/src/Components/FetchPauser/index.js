import { Component } from "react";
import PropTypes from "prop-types";

import { AlertStore } from "Stores/AlertStore";

class FetchPauser extends Component {
  static propTypes = {
    children: PropTypes.any,
    alertStore: PropTypes.instanceOf(AlertStore).isRequired
  };

  componentDidMount() {
    const { alertStore } = this.props;
    alertStore.status.pause();
  }

  componentWillUnmount() {
    const { alertStore } = this.props;
    alertStore.status.resume();
  }

  render() {
    return this.props.children;
  }
}

export { FetchPauser };
