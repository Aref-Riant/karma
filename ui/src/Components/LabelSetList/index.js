import React, { Component } from "react";
import PropTypes from "prop-types";

import { observer } from "mobx-react";
import { observable, action } from "mobx";

import hash from "object-hash";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleLeft } from "@fortawesome/free-solid-svg-icons/faAngleLeft";
import { faAngleRight } from "@fortawesome/free-solid-svg-icons/faAngleRight";
import { faAngleDoubleLeft } from "@fortawesome/free-solid-svg-icons/faAngleDoubleLeft";
import { faAngleDoubleRight } from "@fortawesome/free-solid-svg-icons/faAngleDoubleRight";

import Pagination from "react-js-pagination";

import { AlertStore } from "Stores/AlertStore";
import { StaticLabel } from "Components/Labels/StaticLabel";

// take a list of groups and outputs a list of label sets, this ignores
// the receiver, so we'll end up with only unique alerts
const GroupListToUniqueLabelsList = groups => {
  const alerts = {};
  for (const group of groups) {
    for (const alert of group.alerts) {
      const alertLabels = Object.assign(
        {},
        group.labels,
        group.shared.labels,
        alert.labels
      );
      const alertHash = hash(alertLabels);
      alerts[alertHash] = alertLabels;
    }
  }
  return Object.values(alerts);
};

const LabelSetList = observer(
  class LabelSetList extends Component {
    static propTypes = {
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
      labelsList: PropTypes.arrayOf(PropTypes.object).isRequired
    };

    maxPerPage = 10;

    pagination = observable(
      {
        activePage: 1,
        onPageChange(pageNumber) {
          this.activePage = pageNumber;
        }
      },
      {
        onPageChange: action.bound
      }
    );

    render() {
      const { alertStore, labelsList } = this.props;

      return labelsList.length > 0 ? (
        <div>
          <p className="lead text-center">Affected alerts</p>
          <div>
            <ul className="list-group list-group-flush mb-3">
              {labelsList
                .slice(
                  (this.pagination.activePage - 1) * this.maxPerPage,
                  this.pagination.activePage * this.maxPerPage
                )
                .map(labels => (
                  <li
                    key={hash(labels)}
                    className="list-group-item px-0 pt-2 pb-1"
                  >
                    {Object.entries(labels).map(([name, value]) => (
                      <StaticLabel
                        key={name}
                        alertStore={alertStore}
                        name={name}
                        value={value}
                      />
                    ))}
                  </li>
                ))}
            </ul>
          </div>
          {labelsList.length > this.maxPerPage ? (
            <div className="mt-3">
              <Pagination
                activePage={this.pagination.activePage}
                itemsCountPerPage={this.maxPerPage}
                totalItemsCount={labelsList.length}
                pageRangeDisplayed={5}
                onChange={this.pagination.onPageChange}
                hideFirstLastPages={labelsList.length / this.maxPerPage < 10}
                innerClass="pagination justify-content-center"
                itemClass="page-item"
                linkClass="page-link"
                activeClass="active"
                activeLinkClass="font-weight-bold"
                prevPageText={<FontAwesomeIcon icon={faAngleLeft} />}
                nextPageText={<FontAwesomeIcon icon={faAngleRight} />}
                firstPageText={<FontAwesomeIcon icon={faAngleDoubleLeft} />}
                lastPageText={<FontAwesomeIcon icon={faAngleDoubleRight} />}
              />
            </div>
          ) : null}
        </div>
      ) : (
        <p className="text-muted text-center">No alerts matched</p>
      );
    }
  }
);

export { LabelSetList, GroupListToUniqueLabelsList };
