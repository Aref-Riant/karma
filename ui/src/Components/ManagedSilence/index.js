import React, { Component } from "react";
import PropTypes from "prop-types";

import { observable, action } from "mobx";
import { observer } from "mobx-react";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";

import { APISilence } from "Models/API";
import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore, SilenceTabNames } from "Stores/SilenceFormStore";
import { MountFade } from "Components/Animations/MountFade";
import { SilenceComment } from "./SilenceComment";
import { SilenceDetails } from "./SilenceDetails";

import "./index.scss";

const ManagedSilence = observer(
  class ManagedSilence extends Component {
    static propTypes = {
      cluster: PropTypes.string.isRequired,
      silence: APISilence.isRequired,
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
      silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
      onDidUpdate: PropTypes.func,
      onDeleteModalClose: PropTypes.func
    };

    // store collapse state, by default only silence comment is visible
    // the rest of the silence is hidden until expanded by a click
    collapse = observable(
      {
        value: true,
        toggle() {
          this.value = !this.value;
        }
      },
      { toggle: action.bound }
    );

    getAlertmanager = () =>
      this.props.alertStore.data.upstreams.instances
        .filter(u => u.cluster === this.props.cluster)
        .slice(0, 1)[0];

    onEditSilence = () => {
      const { silenceFormStore, silence } = this.props;

      const alertmanager = this.getAlertmanager();

      silenceFormStore.data.fillFormFromSilence(alertmanager, silence);
      silenceFormStore.data.resetProgress();
      silenceFormStore.tab.setTab(SilenceTabNames.Editor);
      silenceFormStore.toggle.show();
    };

    componentDidUpdate() {
      const { onDidUpdate } = this.props;
      if (onDidUpdate) onDidUpdate();
    }

    render() {
      const {
        cluster,
        silence,
        alertStore,
        silenceFormStore,
        onDeleteModalClose
      } = this.props;

      return (
        <MountFade in={true}>
          <div className="card my-1 components-managed-silence">
            <div className="card-header border-bottom-0">
              <div className="d-flex flex-row justify-content-between mw-100">
                <div className="flex-grow-1 flex-shrink-1 mr-2">
                  <div className="my-1">
                    <SilenceComment
                      silence={silence}
                      collapsed={this.collapse.value}
                    />
                  </div>
                </div>
                <div className="flex-grow-0 flex-shrink-0 mt-auto mb-0">
                  <FontAwesomeIcon
                    icon={this.collapse.value ? faChevronUp : faChevronDown}
                    className="text-muted cursor-pointer"
                    onClick={this.collapse.toggle}
                  />
                </div>
              </div>
            </div>

            {this.collapse.value ? null : (
              <div className="card-body pt-0">
                <SilenceDetails
                  cluster={cluster}
                  silence={silence}
                  alertStore={alertStore}
                  silenceFormStore={silenceFormStore}
                  onEditSilence={this.onEditSilence}
                  onDeleteModalClose={onDeleteModalClose}
                />
              </div>
            )}
          </div>
        </MountFade>
      );
    }
  }
);

export { ManagedSilence };
