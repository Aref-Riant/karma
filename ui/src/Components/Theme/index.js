import React, { Component } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

const ThemeContext = React.createContext();

class BodyTheme extends Component {
  onToggleBodyClass = (isDark) => {
    document.body.classList.toggle("theme-dark", isDark);
    document.body.classList.toggle("theme-light", !isDark);
  };

  componentDidMount() {
    this.onToggleBodyClass(this.context.isDark);
  }

  componentDidUpdate() {
    this.onToggleBodyClass(this.context.isDark);
  }

  render() {
    return (
      <React.Suspense fallback={<Placeholder />}>
        {this.context.isDark ? <DarkTheme /> : <LightTheme />}
      </React.Suspense>
    );
  }
}
BodyTheme.contextType = ThemeContext;

export { BodyTheme, ThemeContext };
