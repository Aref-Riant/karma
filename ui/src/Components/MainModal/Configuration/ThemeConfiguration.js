import React, { Component } from "react";
import PropTypes from "prop-types";

import { action } from "mobx";
import { observer } from "mobx-react";

import Select from "react-select";

import { Settings } from "Stores/Settings";
import { ThemeContext } from "Components/Theme";

const ThemeConfiguration = observer(
  class ThemeConfiguration extends Component {
    static propTypes = {
      settingsStore: PropTypes.instanceOf(Settings).isRequired,
    };
    static contextType = ThemeContext;

    constructor(props) {
      super(props);

      this.validateConfig();
    }

    valueToOption = (val) => {
      const { settingsStore } = this.props;

      return {
        label: settingsStore.themeConfig.options[val].label,
        value: val,
      };
    };

    validateConfig = action(() => {
      const { settingsStore } = this.props;

      if (
        !Object.values(settingsStore.themeConfig.options)
          .map((o) => o.value)
          .includes(settingsStore.themeConfig.config.theme)
      ) {
        settingsStore.themeConfig.config.theme =
          settingsStore.themeConfig.options.auto.value;
      }
    });

    onCollapseChange = action((newValue, actionMeta) => {
      const { settingsStore } = this.props;

      settingsStore.themeConfig.config.theme = newValue.value;
    });

    render() {
      const { settingsStore } = this.props;

      return (
        <div className="form-group mb-2">
          <Select
            styles={this.context.reactSelectStyles}
            classNamePrefix="react-select"
            instanceId="configuration-theme"
            defaultValue={this.valueToOption(
              settingsStore.themeConfig.config.theme
            )}
            options={Object.values(settingsStore.themeConfig.options)}
            onChange={this.onCollapseChange}
            hideSelectedOptions
          />
        </div>
      );
    }
  }
);

export { ThemeConfiguration };
