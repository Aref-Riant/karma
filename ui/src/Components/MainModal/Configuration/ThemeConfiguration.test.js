import React from "react";

import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { Settings } from "Stores/Settings";
import { ThemeContext } from "Components/Theme";
import {
  ReactSelectColors,
  ReactSelectStyles
} from "Components/Theme/ReactSelect";
import { ThemeConfiguration } from "./ThemeConfiguration";

let settingsStore;
beforeEach(() => {
  settingsStore = new Settings();
});

const FakeConfiguration = () => {
  return mount(
    <ThemeContext.Provider
      value={{
        reactSelectStyles: ReactSelectStyles(ReactSelectColors.Light)
      }}
    >
      <ThemeConfiguration settingsStore={settingsStore} />
    </ThemeContext.Provider>
  );
};

describe("<ThemeConfiguration />", () => {
  it("matches snapshot with default values", () => {
    const tree = FakeConfiguration();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("resets stored config to defaults if it is invalid", done => {
    settingsStore.themeConfig.config.theme = "foo";
    const tree = FakeConfiguration();
    const select = tree.find("div.react-select__value-container");
    expect(select.text()).toBe(settingsStore.themeConfig.options.auto.label);
    setTimeout(() => {
      expect(settingsStore.themeConfig.config.theme).toBe(
        settingsStore.themeConfig.options.auto.value
      );
      done();
    }, 200);
  });

  it("rendered correct default value", done => {
    settingsStore.themeConfig.config.theme =
      settingsStore.themeConfig.options.auto.value;
    const tree = FakeConfiguration();
    const select = tree.find("div.react-select__value-container");
    setTimeout(() => {
      expect(select.text()).toBe(settingsStore.themeConfig.options.auto.label);
      done();
    }, 200);
  });

  it("clicking on a label option updates settingsStore", done => {
    const tree = FakeConfiguration();
    tree
      .find("input#react-select-configuration-theme-input")
      .simulate("change", { target: { value: " " } });
    const options = tree.find("div.react-select__option");
    options.at(1).simulate("click");
    setTimeout(() => {
      expect(settingsStore.themeConfig.config.theme).toBe(
        settingsStore.themeConfig.options.dark.value
      );
      done();
    }, 200);
  });
});
