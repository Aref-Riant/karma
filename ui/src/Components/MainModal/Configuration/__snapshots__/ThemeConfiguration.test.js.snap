// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeConfiguration /> matches snapshot with default values 1`] = `
"
<div class=\\"form-group mb-0\\">
  <div class=\\"form-check form-check-inline\\">
    <span class=\\"custom-control custom-switch\\">
      <input id=\\"configuration-theme\\"
             class=\\"custom-control-input\\"
             type=\\"checkbox\\"
             value
      >
      <label class=\\"custom-control-label cursor-pointer mr-3\\"
             for=\\"configuration-theme\\"
      >
        Enable dark mode
      </label>
      <span class=\\"ml-5 badge badge-danger align-text-bottom\\">
        Experimental
      </span>
    </span>
  </div>
</div>
"
`;
